spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/metrics/EurekaInstanceMetricsBinder.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2013-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.netflix.eureka.server.metrics;
+
+import java.util.List;
+import java.util.Objects;
+
+import com.netflix.eureka.registry.PeerAwareInstanceRegistry;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
+/**
+ * @author heowc
+ */
+public class EurekaInstanceMetricsBinder implements MeterBinder {
+
+	private final PeerAwareInstanceRegistry instanceRegistry;
+
+	EurekaInstanceMetricsBinder(PeerAwareInstanceRegistry instanceRegistry) {
+		this.instanceRegistry = Objects.requireNonNull(instanceRegistry);
+	}
+
+	@Override
+	public void bindTo(MeterRegistry meterRegistry) {
```

Comment:
This will be called once, when the binder is registered and never again.
How do you expect registrations to be updated?

I _think_ you can simulate this in your tests by doing something like this:
```java
InstanceInfo instanceInfo = getInstanceInfo(...);
binder.bindTo(meterRegistry);
instanceRegistry.register(instanceInfo, false);
assert... // You will see zero instances but there should be one
```

There are multiple ways to resolve this and I can help but first I want to make sure we are on the same page and I'm not missing something.
(We can use a `ScheduledTaskExecutor`, the `Gauge` itself to refresh the list of instances and also a `MultiGauge` can be helpful in some cases.)

Reply:
No, I think it's not been done on purpose. We should probably create these metrics during events related to instance lifetime, such as `EurekaInstanceRegisteredEvent` and `EurekaInstanceCancelledEvent`.

Reply:
or use `Gauge` to update results with information from new `application.getInstances()` periodically.

Reply:
>We should probably create these metrics during events related to instance lifetime, such as EurekaInstanceRegisteredEvent and EurekaInstanceCancelledEvent.

This seem to be the most promising one to me.
@heowc do you want to look into this? I think if you listen to these events, you can use `MutiGauge` to do the recording. This might play well with the user-defined tags/aggregation we discussed above: if you pass the list of instances, the user-defined logic can return a list of `MultiGauge.Row` instances that define the tags and the number of instances too. Please let me know if you need help/have questions. You can find the docs here: [Multi Gauge](https://docs.micrometer.io/micrometer/reference/concepts/gauges.html#_multi_gauge) and there is some sample code here: [`MultiGaugeSample`](https://github.com/micrometer-metrics/micrometer/blob/main/samples/micrometer-samples-core/src/main/java/io/micrometer/core/samples/MultiGaugeSample.java).

Reply:
Ooops..! Thank you.

Let me look into this.

Reply:
While examining `EurekaInstanceRegisteredEvent` and `EurekaInstanceCancelledEvent`, I found something somewhat strange. I've written a related PR, please check it. #4274  🤔

Reply:
@OlgaMaciaszek @jonatan-ivanov 

I made a lot of changes to use `MultiGuage`.

- I changed it to listen for events like `EurekaInstanceXXXEvent`.
- I changed to using `TagsProvider` to group and count things with equivalent tags.
- However, since `status` is required, I always add it independently of the `TagsProvider`.

If you have any other good ideas, feel free to let me know 🙏